File: static_language_c_plus_plus/sudoku/grid.cpp
#include "grid.h"

Grid::Grid(int grid_size, int box_size)
    :   _grid_size(grid_size),
        _box_size(box_size)
{
    // Initialize grid
    _data.resize(grid_size, std::vector<int>(grid_size, 0));
}

std::vector<int> Grid::getColumn(int column) const
{
    std::vector<int> column_data(_grid_size, 0);

    // 1. Check if row is valid
    if ((unsigned int)column >= _grid_size)
    {
        return column_data;
    }

    // 2. Construct return vector
    for (int row = 0; row < _grid_size; ++row)
    {
        column_data[row] = _data[row][column];
    }
    return column_data;
}

std::vector<int> Grid::getBox(int row, int column) const
{
    std::vector<int> box_data(_box_size * _box_size, 0);

    // 1. Check if row and column are valid
    if ((unsigned int)row >= _grid_size || (unsigned int)column >= _grid_size)
    {
        return box_data;
    }

    // 2. Construct box data
    int box_begin_row = row / _box_size * _box_size,
        box_begin_column = column / _box_size * _box_size;
    for (int r = 0; r < _box_size; ++r)
    {
        for (int c = 0; c < _box_size; ++c)
        {
            box_data[r * _box_size + c] = _data[r + box_begin_row][c + box_begin_column];
        }
    }
    return box_data;
}

void Grid::resetData()
{
    _data.clear();
    _data.resize(_grid_size, std::vector<int>(_grid_size, 0));
}

File: dynamic_language_python/sudoku/grid.py
import numpy as np

class Grid:
    def __init__(self, grid_size, box_size):
        self._grid_size = grid_size
        self._box_size = box_size
        self._data = np.zeros((grid_size, grid_size), dtype=int)

    # 获得第row行数据
    def getRow(self, row):
        if row >= self._grid_size or row < 0: 
            return np.array()

        return self._data[row, :]

    # 获得第column列数据
    def getColumn(self, column):
        if column >= self._grid_size or column < 0:
            return np.array()
        
        return self._data[:, column]
    
    # 获取第row行第column列所在box的数据
    def getBox(self, row, column):
        if row >= self._grid_size or row < 0 or         \
            column >= self._grid_size or column < 0:
            return np.array()

        box_begin_index = row // self._box_size
        box_end_index = column // self._box_size
        return self._data[box_begin_index : box_begin_index + self._box_size,
                          box_end_index   : box_end_index + self._box_size].flatten()

    # 获取第row行第column列数字
    def at(self, row, column):
        if row >= self._grid_size or row < 0 or         \
            column >= self._grid_size or column < 0:
            return 0

        return self._data[row][column]

    # 返回网格size
    def gridSize(self):
        return self._grid_size

    # 重置网格数据
    def _resetData(self):
        self._data = np.zeros((self._grid_size, self._grid_size), dtype=int)